### PR TITLE
Fixing #12161: rename Bool.leb into Bool.le

### LINKED ISSUE #12161: [stdlib] [bool] Unconventional naming for order relation Bool.leb
The order relation on `bool` is named `leb : bool -> bool -> Prop` while such a name usually suggests `bool` as output type: `leb : bool -> bool -> bool`.
It would be more uniform with the rest of the standard library to consider naming it `le` (see for example `Nat.leb` and `Nat.le`).

(this comes from a discussion in #12008)

### PR DESCRIPTION
**Pull request by olaure01**
This PR splits from #12008 keeping the renaming part only.
See #12008 for the beginning of the discussion involving @herbelin and @JasonGross.

**Kind:** feature

Fixes / closes #12161 

- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by Zimmi48** [comment]
There are multiple CI failures, and at least two are related to omega not being able to solve a goal anymore. This is likely because `omega` does not use the `Coqlib` / `Register` mechanism to find the constants it knows about.

**Comment by olaure01** [comment]
@Zimmi48 I see 1 because of `omega` (according to #11976, probably `lia` should be tried) and 5 because of scope troubles probably solvable by qualifying some `le` into `Nat.le`, but I may be wrong.
The first question is: is there an agreement on this name change? (no objection yet, including in the discussion started in #12008 previously).
The second is: if so, what could I do to move on?

**Comment by Zimmi48** [comment]
>  The second is: if so, what could I do to move on?

If there's an agreement, then overlays will need to be prepared (and obviously a changelog entry documenting the change). See https://github.com/coq/coq/blob/master/dev/ci/README-developers.md. Hopefully, the required fixes will be small and easy to make backward-compatible.

**Comment by olaure01** [comment]
Apparently all the overlays are working. The remaining CI problems with lambda-rust and perennial come in fact from std++, for which a patch is submitted as a [merge request](https://gitlab.mpi-sws.org/iris/stdpp/-/merge_requests/153) on the corresponding gitlab.

**Comment by JasonGross** [comment]
Note that you should make PRs for the overlays and we would merge them *before* this PR is merged

**Comment by olaure01** [comment]
@JasonGross Agreed. Here is my current state of mind regarding this PR:

- It is not an easy one (at least the most difficult one I have been involved into so far: large impact, first time I am facing CI and overlays, etc) since it has an impact on many developments.
- If I have seen nobody going against it so far, I am prepared to see some counter arguments and I do not want to bother too many people with useless PR on other projects if at some point the current one is dismissed (no approving review yet).
- I would like to be sure the patch for std++ (leading to the lambda-rust and perennial problems in CI) is integrated and works well (ongoing [discussion there](https://gitlab.mpi-sws.org/iris/stdpp/-/merge_requests/153)).
- There are potential interactions with #12008: since after it ``Bool.ltb`` will have to be renamed into ``Bool.lt`` (generating conflicts with ``Nat.lt`` as well). #12008 being much harmless, I am prepared to see it merged before the current one. If this happens, it looks reasonable to updated the present PR with all renamings and to avoid multiplying external PRs for the additional overlays to come.

But maybe I am wrong and we should go on in parallel and I can make the 4 external PRs: they are ready on my forks. Any advice welcome.

**Comment by JasonGross** [comment]
I'm happy to merge the PR(s) on fiat-parsers, as long as they are backwards-compatible.  I also support the renaming of leb to le that's done in this PR; I think it's a good idea.  (As far as I can tell, no one is against the renaming, just concerned about compatibility problems.)

**Comment by Zimmi48** [comment]
I also support the renaming. The compatibility impact seems medium (not that big for such a core definition). AFICT, all the prepared overlays are backward-compatible. I also agree with your approach of delaying the submission of the overlays to get some support for the PR first, but sometimes submitting the overlays will be the only way to actually get feedback. Finally, if this renaming is to be done, it should be done in 8.12. Remember that the branching date is May 15th, so the PR must be merged before. *In particular, I wouldn't like to see a release containing `Bool.ltb` to be later renamed into `Bool.lt` in a further release.*

**Comment by Zimmi48** [comment]
In the related std++ MR, @RalfJung said:

>If Coq's module system would provide a way to make sure le is always qualified, I would be inclined to agree or at least make the experiment. But I think such an approach is hopeless to enforce with the current module system.
But also, we now already have a large body of code that uses le without qualification...

Thanks for the great idea @RalfJung! It happens to be the case that Coq's module system does support such a feature. You just have to use the `local` attribute / `Local` prefix on the thing that you want to be always qualified. Example:

```coq
Module foo.
#[ local ] Definition bar := 0.
End foo.

Import foo.

Fail Check bar.
Check foo.bar.
```

Therefore, this opens the path to a solution for this PR that will strongly (AFAICT) reduce the compatibility breakage (and sorry @olaure01 for not having thought of this earlier): you just have to make the renamed definition of `le` and `lt` in the `Bool` module local. Example:

```coq
Module Nat.
Definition bar := 1.
End Nat.

Module Bool.
#[ local ] Definition bar := true.
End Bool.

Import Nat.
Import Bool.

Print bar. (* bar = 1 : nat *)
```

Making `Nat.le` also always be qualified would be nice as well but a way bigger breaking change, and we would need to find a way of having it qualified as `Nat.le` and not as `PeanoNat.le`... so definitely not for today!

**Comment by RalfJung** [comment]
@Zimmi48 interesting, I did not know `Local` can do that!
Is there a difference between `Local Definition` and `#[local] Definition`?

**Comment by Zimmi48** [comment]
>  Is there a difference between `Local Definition` and `#[local] Definition`?

Nope, as better explained in the freshly improved documentation on attributes: https://coq.github.io/doc/master/refman/language/core/basic.html#attributes

**Comment by RalfJung** [comment]
FWIW, we also landed the changes now to make std++ compatible with the original proposal here, so if you re-run CI it should work.

**Comment by olaure01** [comment]
New version based on local definitions. CI ok with no need for overlays.

**Comment by olaure01** [comment]
@herbelin we finally have a reasonable solution thanks to @Zimmi48

File: doc/changelog/10-standard-library/12162-bool-leb.rst
Comment (Zimmi48) [inline doc/changelog/10-standard-library/12162-bool-leb.rst:2]: This could also be "**Deprecated:** `Bool.leb` in favor of `Bool.le`". With the rest virtually unchanged.

**Comment by Zimmi48** [comment]
@herbelin or @anton-trunov This is ready and just needs an assignee.

**Review by anton-trunov — APPROVED** [review]

**Pull request merged**
